store/sqlstore/channel_store.go

```diff
@@ -1812,6 +1812,69 @@ func (s SqlChannelStore) UpdateLastViewedAt(channelIds []string, userId string)
 	return times, nil
 }
 
+// CountUnreadSince gives the number of posts that a channel has since some date, this doesn't follow any threads.
+// this allows us to know how many posts have bean read already.
+func (s SqlChannelStore) CountUnreadSince(channelID string, since int64) (int64, *model.AppError) {
+	countUnreadQuery := `
+	SELECT count(*)
+	FROM Posts
+	WHERE
+		ChannelId = :channelID
+		AND CreateAt < :createAt
+		AND Type = ""
+		AND Deleteat = 0
+	`
+	countParams := map[string]interface{}{
+		"channelID": channelID,
+		"createAt":  since,
+	}
+
+	unread, err := s.GetReplica().SelectInt(countUnreadQuery, countParams)
+	if err != nil {
+		return 0, model.NewAppError("SqlChannelStore.CountUnreadSince", "store.sql_channel.CountUnreadSince.app_error", countParams, fmt.Sprintf("Error while counting unreads for %s since %d", channelID, since), http.StatusInternalServerError)
+	}
+	return unread, nil
+}
+
+// UpdateLastViewedAtPost sets a channel as unread for a user at the time of the post selected and update the MentionCount
+// it returns a channelunread so redux can update the apps easily.
+func (s SqlChannelStore) UpdateLastViewedAtPost(unreadPost model.Post, userID string, mentionCount int) (*model.ChannelUnread, *model.AppError) {
+
+	unread, appErr := s.CountUnreadSince(unreadPost.ChannelId, unreadPost.CreateAt)
+	if appErr != nil {
+		return nil, appErr
+	}
+
+	params := map[string]interface{}{
+		"mentions":     mentionCount,
+		"unreadCount":  unread,
+		"lastViewedAt": unreadPost.CreateAt,
+		"userID":       userID,
+		"channelID":    unreadPost.ChannelId,
+	}
+
+	// msg count uses the value from channels to prevent counting on older channels where no. of messages can be high.
+	// we only count the unread which will be a lot less in 99% cases
+	setUnreadQuery := `
+	UPDATE
+		ChannelMembers
+	SET
+		MentionCount = :mentions,
+		MsgCount = (SELECT TotalMsgCount FROM Channels WHERE ID = :channelID) - :unreadCount,
+		LastViewedAt = :lastViewedAt,
+		LastUpdateAt = :lastViewedAt
```

Comment:
I think we might want to set LastUpdateAt to the current time. I think it's meant to keep track of the last time the row changed regardless of what it changed to

Reply:
Sure! I thought this was just meant for posts (creation, edition, deletion), as this wouldn't change the channel, but just when it was watched.